File: validation/differential/corpus/cognitive/typescript/constructs.ts
// TypeScript cognitive-complexity micro-corpus (G1-A).
//
// Every callable pins named rules from FROZEN_RULE_TABLE.md. Expected values in
// expectations.json are authored FROM THAT TABLE and from nothing else.
//
// The JSX short-circuit divergence lives in constructs.tsx: `.ts` and `.tsx` are
// different parsers to ArchLens, and JSX is only legal in the latter.

type Maybe<T> = T | null;

export function trivial(a: number): number {
  return a;
}

export function plainIf(a: boolean): number {
  if (a) {
    return 1;
  }
  return 0;
}

export function ifElseIfElse(a: number): number {
  if (a === 1) {
    return 1;
  } else if (a === 2) {
    return 2;
  } else {
    return 3;
  }
}

export function nestedIf(a: boolean, b: boolean): number {
  if (a) {
    if (b) {
      return 1;
    }
  }
  return 0;
}

export function loops(xs: number[]): number {
  let total = 0;
  for (const x of xs) {
    while (total < x) {
      total += 1;
    }
  }
  return total;
}

export function switchWhole(a: number): number {
  switch (a) {
    case 1:
      return 1;
    case 2:
      return 2;
    default:
      return 3;
  }
}

export function exceptionHandling(a: string): number {
  try {
    return Number(a);
  } catch (error) {
    return 0;
  } finally {
    if (a) {
      String(a);
    }
  }
}

export function ternary(a: boolean): number {
  return a ? 1 : 0;
}

export function boolSameOperator(a: boolean, b: boolean, c: boolean): number {
  if (a && b && c) {
    return 1;
  }
  return 0;
}

export function boolMixedOperators(a: boolean, b: boolean, c: boolean): number {
  if (a && b || c) {
    return 1;
  }
  return 0;
}

export function nullishSequence(a: Maybe<number>, b: Maybe<number>): number {
  return (a ?? b) ?? 1;
}

export function optionalChaining(a: { b?: { c?: number } }): number | undefined {
  return a?.b?.c;
}

export function bitwiseNotCounted(a: number, b: number): number {
  if ((a & b) !== 0) {
    return 1;
  }
  return 0;
}

export function labelledFlow(xs: number[]): number {
  outer: for (const x of xs) {
    for (const y of xs) {
      if (x > y) {
        break outer;
      }
      continue outer;
    }
  }
  return 0;
}

export function unlabelledJump(xs: number[]): number {
  for (const x of xs) {
    break;
  }
  return 0;
}

export function recursive(n: number): number {
  if (n <= 0) {
    return 0;
  }
  return recursive(n - 1);
}

export function callbackExcluded(xs: number[]): number[] {
  return xs.map((x) => {
    if (x > 0) {
      return x;
    }
    return 0;
  });
}

export function typeOnlyConstructs(a: unknown): number {
  const widened = a as Maybe<number>;
  const forced = widened!;
  return forced;
}

export function overloaded(a: number): number;
export function overloaded(a: string): number;
export function overloaded(a: number | string): number {
  return typeof a === "number" ? a : a.length;
}

export class Widget {
  render(a: boolean): number {
    if (a) {
      return 1;
    }
    return 0;
  }
}

File: validation/differential/corpus/cognitive/java/Constructs.java
// Java cognitive-complexity micro-corpus (G1-A).
//
// Every callable pins named rules from FROZEN_RULE_TABLE.md. Expected values in
// expectations.json are authored FROM THAT TABLE and from nothing else.
//
// JDK 21 `case X when g` lives in ConstructsGuarded.java, not here: the pinned
// reference JDK is 17 and cannot compile it. That is a toolchain limit, not a
// limit on whether the rule is expressible in Java.

class Constructs {

    int trivial(int a) {
        return a;
    }

    int plainIf(boolean a) {
        if (a) {
            return 1;
        }
        return 0;
    }

    int ifElseIfElse(int a) {
        if (a == 1) {
            return 1;
        } else if (a == 2) {
            return 2;
        } else {
            return 3;
        }
    }

    int nestedIf(boolean a, boolean b) {
        if (a) {
            if (b) {
                return 1;
            }
        }
        return 0;
    }

    int loops(int[] xs) {
        int total = 0;
        for (int x : xs) {
            while (x > 0) {
                x -= 1;
                total += 1;
            }
        }
        return total;
    }

    int switchWhole(int a) {
        switch (a) {
            case 1:
                return 1;
            case 2:
                return 2;
            default:
                return 3;
        }
    }

    int switchExpression(int a) {
        return switch (a) {
            case 1 -> 1;
            case 2 -> 2;
            default -> 3;
        };
    }

    int exceptionHandling(String a) {
        try {
            return Integer.parseInt(a);
        } catch (NumberFormatException e) {
            return 0;
        } catch (RuntimeException e) {
            return -1;
        } finally {
            if (a != null) {
                a.length();
            }
        }
    }

    int multiCatch(String a) {
        try {
            return Integer.parseInt(a);
        } catch (NumberFormatException | NullPointerException e) {
            return 0;
        }
    }

    int ternary(boolean a) {
        return a ? 1 : 0;
    }

    int boolSameOperator(boolean a, boolean b, boolean c) {
        if (a && b && c) {
            return 1;
        }
        return 0;
    }

    int boolMixedOperators(boolean a, boolean b, boolean c) {
        if (a && b || c) {
            return 1;
        }
        return 0;
    }

    int bitwiseNotCounted(int a, int b) {
        if ((a & b) != 0) {
            return 1;
        }
        return 0;
    }

    int labelledFlow(int[] xs) {
        outer:
        for (int x : xs) {
            for (int y : xs) {
                if (x > y) {
                    break outer;
                }
                continue outer;
            }
        }
        return 0;
    }

    int unlabelledJump(int[] xs) {
        for (int x : xs) {
            break;
        }
        return 0;
    }

    int recursive(int n) {
        if (n <= 0) {
            return 0;
        }
        return recursive(n - 1);
    }

    int synchronizedBlock(boolean a) {
        synchronized (this) {
            if (a) {
                return 1;
            }
        }
        return 0;
    }

    Runnable lambdaExcluded(boolean a) {
        return () -> {
            if (a) {
                System.out.print(1);
            }
        };
    }

    Runnable anonymousClassExcluded(boolean a) {
        return new Runnable() {
            @Override
            public void run() {
                if (a) {
                    System.out.print(1);
                }
            }
        };
    }
}
